arch/arm/core/cortex_m/mpu/arm_mpu.c

```diff
@@ -78,251 +126,138 @@ void arm_core_mpu_disable(void)
+	u8_t i;
+	u8_t reg_index = _get_num_regions();
 
-/**
- * This internal function is utilized by the MPU driver to parse the intent
- * type (i.e. THREAD_STACK_REGION) and return the correct region index.
- */
-static inline u32_t _get_region_index_by_type(u32_t type)
-{
-	u32_t region_index;
+	for (i = static_regions_num; i < _get_num_regions(); i++) {
+		if (!_is_enabled_region(i)) {
+			continue;
+		}
 
-	__ASSERT(type < THREAD_MPU_REGION_LAST,
-		 "unsupported region type");
+		u32_t base = _mpu_region_get_base(i);
 
-	region_index = mpu_config.num_regions + type;
+		if (base != partition->start) {
+			continue;
+		}
 
-	__ASSERT(region_index < _get_num_regions(),
-		 "out of MPU regions, requested %u max is %u",
-		 region_index, _get_num_regions() - 1);
+		u32_t size = _mpu_region_get_size(i);
 
-	return region_index;
-}
+		if (size != partition->size) {
+			continue;
+		}
 
-/**
- * This internal function disables a given MPU region.
- */
-static inline void _disable_region(u32_t r_index)
-{
-	/* Attempting to configure MPU_RNR with an invalid
-	 * region number has unpredictable behavior. Therefore
-	 * we add a check before disabling the requested MPU
-	 * region.
-	 */
-	__ASSERT(r_index < _get_num_regions(),
-		"Index 0x%x out-of-bound (supported regions: 0x%x)\n",
-		r_index,
-		_get_num_regions());
-	LOG_DBG("disable region 0x%x", r_index);
-	/* Disable region */
-	ARM_MPU_ClrRegion(r_index);
+		/* Region found */
+		reg_index = i;
+		break;
+	}
+	__ASSERT(reg_index != _get_num_regions(),
+		"Memory domain partition not found\n");
+
+	/* Modify the permissions */
+	partition->attr = *new_attr;
+	_mpu_configure_region(reg_index, partition);
 }
 
 /**
- * @brief configure the base address and size for an MPU region
- *
- * @param   type    MPU region type
- * @param   base    base address in RAM
- * @param   size    size of the region
+ * @brief get the maximum number of free regions for memory domain partitions
  */
-void arm_core_mpu_configure(u8_t type, u32_t base, u32_t size)
+int arm_core_mpu_get_max_domain_partition_regions(void)
 {
-	struct arm_mpu_region region_conf;
-
-	LOG_DBG("Region info: 0x%x 0x%x", base, size);
-	u32_t region_index = _get_region_index_by_type(type);
-
-	if (_get_region_attr_by_type(&region_conf.attr, type, base, size)) {
-		return;
-	}
-	region_conf.base = base;
-
-	if (region_index >= _get_num_regions()) {
-		return;
-	}
-
-	_region_init(region_index, &region_conf);
-}
+	int available_regions_num =	_get_num_regions() - static_regions_num;
 
-#if defined(CONFIG_USERSPACE)
-void arm_core_mpu_configure_user_context(struct k_thread *thread)
-{
-	if (!thread->arch.priv_stack_start) {
-		_disable_region(_get_region_index_by_type(
-			THREAD_STACK_REGION));
-		return;
-	}
+	/* Additional regions required for the thread stack,
+	 * depending on whether the MPU architecture requires
+	 * non overlapping regions.
+	 */
+#if defined(CONFIG_MPU_REQUIRES_NON_OVERLAPPING_REGIONS)
```

Comment:
We should not take under account any thread stack and other SW configuration. This layer should get this information from kernel (of course CONFIG_MPU_REQUIRES_NON_OVERLAPPING_REGIONS as HW configuration should be handled here).

Reply:
Yeah, I see what you mean:
You propose implement the function in arm_core_mpu.c and let the driver (arm_mpu.c, nxp_mpu.c) only return the number of available (HW) regions indices.

Do I understand corretly, @pizi-nordic ?

Reply:
Yes.

Reply:
Addressed now, thanks.